### PR TITLE
Do not use stream encoders for pipe communication

### PR DESCRIPTION
**Pull request by crosbymichael**
Marshall the raw objects for the sync pipes so that no new line chars
are left behind in the pipe causing errors.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

### CODE REVIEW

File: libcontainer/utils/utils.go [hunk 1]
```diff
@@ -36,10 +39,27 @@ func ResolveRootfs(uncleanRootfs string) (string, error) {
...
+	if err != nil {
+		return err
+	}
+	buf := bytes.NewBuffer(data)
+	n, err := buf.WriteTo(w)
+	if err != nil {
+		return err
+	}
+	if int(n) != len(data) {
+		return fmt.Errorf("object not written")
```
Comment (dqminh) [inline libcontainer/utils/utils.go]: would be nice to have the returned length and expected value here, something like: `fmt.Errorf("object not written, got(%d), expected(%d)", n, len(data))`

File: libcontainer/utils/utils.go [hunk 2]
```diff
@@ -36,10 +39,27 @@ func ResolveRootfs(uncleanRootfs string) (string, error) {
...
 	return status.ExitStatus()
 }
+
+// WriteJSON writes the provided struct v to w using standard json marshaling
+func WriteJSON(w io.Writer, v interface{}) error {
+	data, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	buf := bytes.NewBuffer(data)
```
Comment (LK4D4) [inline libcontainer/utils/utils.go]: What point of this instead of just w.Write?

**Comment by LK4D4** [comment]
I think having a separate pipe for sync would significantly simplify the code. But for now, this code is fine.

**Comment by crosbymichael** [comment]
@LK4D4 @dqminh fixed

File: libcontainer/utils/utils.go [hunk 3]
```diff
@@ -36,10 +38,26 @@ func ResolveRootfs(uncleanRootfs string) (string, error) {
...
+	data, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	n, err := w.Write(data)
+	if err != nil {
+		return err
+	}
+	if l := len(data); int(n) != l {
+		return fmt.Errorf("short write of %d compared to %s", n, l)
```
Comment (mrunalp) [inline libcontainer/utils/utils.go]: %s --> %d

Comment (crosbymichael) [inline libcontainer/utils/utils.go]: it was a typo!

Comment (mrunalp) [inline libcontainer/utils/utils.go]: s and d are so close :)

File: libcontainer/utils/utils.go [hunk 4]
```diff
@@ -36,10 +38,26 @@ func ResolveRootfs(uncleanRootfs string) (string, error) {
...
+	data, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	n, err := w.Write(data)
+	if err != nil {
+		return err
+	}
+	if l := len(data); int(n) != l {
+		return fmt.Errorf("short write of %d compared to %d", n, l)
```
Comment (LK4D4) [inline libcontainer/utils/utils.go]: @dqminh I'm pretty sure that `Write` will return `io.ShortWrite` before this line :) but it's ok I think

**Comment by LK4D4** [comment]
LGTM

File: libcontainer/utils/utils.go (hunk 4 above)
Comment (dqminh) [inline libcontainer/utils/utils.go]: yah thats true ( `Write must return a non-nil error if it returns n < len(p)`), so i dont think we need this check

**Comment by mrunalp** [comment]
LGTM. @crosbymichael Not sure if we want to wait for the docker testing results to come back or not before merging.

**Comment by csfrancis** [comment]
Tested the patch provided in https://github.com/opencontainers/runc/pull/508#issuecomment-175187605 against docker 1.9.1 in production and it resolves our issue. Thanks!

**Comment by crosbymichael** [comment]
@csfrancis awesome!  thanks for your help

**Comment by mrunalp** [comment]
Nice!

**Comment by mrunalp** [comment]
I have created https://github.com/docker/docker/pull/19752 to bump the version in docker

**Pull request merged**